Userland/Libraries/LibGfx/ImageFormats/PNGLoader.cpp

```diff
@@ -1,1453 +1,298 @@
-    if (!context.has_seen_actl_chunk_before_idat)
-        return {}; // Ignore if it's not a valid animated png
-
-    if (data.size() != sizeof(fcTL_Chunk))
-        return Error::from_string_literal("fcTL chunk has an abnormal size");
-
-    auto const& fcTL = *bit_cast<fcTL_Chunk* const>(data.data());
-    if (fcTL.sequence_number != context.animation_next_expected_seq)
-        return Error::from_string_literal("Unexpected sequence number");
-
-    context.animation_next_expected_seq++;
-
-    if (fcTL.width == 0 || fcTL.height == 0)
-        return Error::from_string_literal("width and height must be greater than zero in fcTL chunk");
-
-    Checked<int> left { static_cast<int>(fcTL.x_offset) };
-    Checked<int> top { static_cast<int>(fcTL.y_offset) };
-    Checked<int> width { static_cast<int>(fcTL.width) };
-    Checked<int> height { static_cast<int>(fcTL.height) };
-    auto right = left + width;
-    auto bottom = top + height;
-    if (left < 0 || width <= 0 || right.has_overflow() || right > context.width)
-        return Error::from_string_literal("Invalid x_offset value in fcTL chunk");
-    if (top < 0 || height <= 0 || bottom.has_overflow() || bottom > context.height)
-        return Error::from_string_literal("Invalid y_offset value in fcTL chunk");
-
-    bool is_first_animation_frame = context.animation_frames.is_empty();
-    if (!is_first_animation_frame)
-        context.last_completed_animation_frame_index = context.animation_frames.size() - 1;
-
-    context.animation_frames.append({ fcTL });
-
-    if (!context.has_seen_idat_chunk && is_first_animation_frame)
-        context.is_first_idat_part_of_animation = true;
-    return {};
+    return rendered_bitmap;
 }
 
-static ErrorOr<void> process_fdAT(ReadonlyBytes data, PNGLoadingContext& context)
+ErrorOr<size_t> PNGLoadingContext::read_frames(png_structp png_ptr, png_infop info_ptr)
 {
-    // https://www.w3.org/TR/png/#fdAT-chunk
+    if (png_get_acTL(png_ptr, info_ptr, &frame_count, &loop_count)) {
+        // acTL chunk present: This is an APNG.
 
-    if (data.size() <= 4)
-        return Error::from_string_literal("fdAT chunk has an abnormal size");
+        png_set_acTL(png_ptr, info_ptr, frame_count, loop_count);
 
-    u32 sequence_number = *bit_cast<NetworkOrdered<u32> const*>(data.data());
-    if (sequence_number != context.animation_next_expected_seq)
-        return Error::from_string_literal("Unexpected sequence number");
-    context.animation_next_expected_seq++;
+        for (size_t frame_index = 0; frame_index < frame_count; ++frame_index) {
+            png_read_frame_head(png_ptr, info_ptr);
+            u32 width = 0;
+            u32 height = 0;
+            u32 x = 0;
+            u32 y = 0;
+            u16 delay_num = 0;
+            u16 delay_den = 0;
+            u8 dispose_op = 0;
+            u8 blend_op = 0;
+            png_get_next_frame_fcTL(png_ptr, info_ptr, &width, &height, &x, &y, &delay_num, &delay_den, &dispose_op, &blend_op);
```

Comment:
Do you need to check `png_get_valid(png_ptr, info_ptr, PNG_INFO_fcTL)` before calling this?

Reply:
I think you might be right! Let's do it.